server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java

```diff
@@ -1106,15 +1108,16 @@ public String getOffsetFieldName() {
     }
 
     protected void parseCreateField(DocumentParserContext context) throws IOException {
-        String value = context.parser().textOrNull();
-        if (value == null) {
-            value = fieldType().nullValue;
+        var value = context.parser().optimizedTextOrNull();
+
+        if (value == null && fieldType().nullValue != null) {
+            value = new Text(fieldType().nullValue);
         }
 
         boolean indexed = indexValue(context, value);
         if (offsetsFieldName != null && context.isImmediateParentAnArray() && context.canAddIgnoredField()) {
             if (indexed) {
-                context.getOffSetContext().recordOffset(offsetsFieldName, value);
+                context.getOffSetContext().recordOffset(offsetsFieldName, value.string());
```

Comment:
I think there is no need to convert to string here? `UTF8Bytes` already implements `Comparable` and the recordOffset method isn't bound to string, the provided value just needs to implement `Comparable`. This should save another unneeded conversation to java string?

Reply:
Makes sense to me. 

While looking into this I found an issue with the implementation of `UTF8Bytes#compareTo`. It was delegating to `ByteBuffer#compareTo`, which was performing signed comparison. This was incorrect, and it needs to do unsigned comparison to perform correct lexicographical ordering of encoded UTF8 values. So I also fixed that.